File: 24-adv-ds1/2-union-find/main.cpp
// Union Find - C++ example

#include <iostream>
#include <vector>

class UnionFind {
public:
    std::vector<int> parent, rank;
    UnionFind(int size) {
        parent.resize(size);
        rank.resize(size, 0);
        for(int i = 0; i < size; i++)
            parent[i] = i;
    }

    int find(int x) {
        if(parent[x] != x)
            parent[x] = find(parent[x]);  // Path compression
        return parent[x];
    }

    void unionSet(int x, int y) {
        int rootX = find(x);
        int rootY = find(y);
        if(rootX != rootY) {
            if(rank[rootX] > rank[rootY])
                parent[rootY] = rootX;
            else if(rank[rootX] < rank[rootY])
                parent[rootX] = rootY;
            else {
                parent[rootY] = rootX;
                rank[rootX]++;
            }
        }
    }
};

int main() {
    UnionFind uf(5);
    uf.unionSet(0,1);
    uf.unionSet(1,2);
    std::cout << (uf.find(0) == uf.find(2)) << std::endl;  // Output: 1 (true)
    std::cout << (uf.find(3) == uf.find(2)) << std::endl;  // Output: 0 (false)
    return 0;
}

File: 24-adv-ds1/2-union-find/main.py
# Union Find - Python example

class UnionFind:
    def __init__(self, size):
        self.parent = list(range(size))
        self.rank = [0] * size

    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])  # Path compression
        return self.parent[x]

    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)
        if rootX != rootY:
            if self.rank[rootX] > self.rank[rootY]:
                self.parent[rootY] = rootX
            elif self.rank[rootX] < self.rank[rootY]:
                self.parent[rootX] = rootY
            else:
                self.parent[rootY] = rootX
                self.rank[rootX] += 1

if __name__ == "__main__":
    uf = UnionFind(5)
    uf.union(0, 1)
    uf.union(1, 2)
    print(uf.find(0) == uf.find(2))  # Output: True
    print(uf.find(3) == uf.find(2))  # Output: False
